machinist/server.go

```diff
@@ -0,0 +1,258 @@
+package machinist
+
+import (
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"github.com/enfabrica/enkit/lib/token"
+	"github.com/enfabrica/enkit/machinist/rpc/machinist"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"log"
+	"net"
+)
+
+func NewServer(req *ServerRequest, modifiers ...ServerModifier) (*Server, error) {
+	s := &Server{
+		Encoder: req.Encoder,
+	}
+	for _, mod := range modifiers {
+		if err := mod(s); err != nil {
+			return nil, err
+		}
+	}
+	return s, nil
+}
+
+func NewServerRequest() *ServerRequest {
+	return &ServerRequest{}
+}
+
+func (rs *ServerRequest) WithNetListener(l *net.Listener) *ServerRequest {
+	rs.Listener = l
+	return rs
+}
+
+type Server struct {
+	Encoder            token.BinaryEncoder
+	NetListener        net.Listener
+	NetAddr            *net.TCPAddr
+	RunningServer      *grpc.Server
+	GenerateServerCert func() (*x509.Certificate, []byte, *rsa.PrivateKey, error)
+	FetchTlSConfig     func() (*tls.Config, error)
+	CAPem              string
+}
+
+func (s *Server) Start() error {
+	tlsConfig, err := s.FetchTlSConfig()
+	if err != nil {
+		return err
+	}
+	ca := credentials.NewTLS(tlsConfig)
+	grpcServer := grpc.NewServer(grpc.Creds(ca))
+	machinist.RegisterControllerServer(grpcServer, s)
+	s.RunningServer = grpcServer
+	return grpcServer.Serve(s.NetListener)
+}
+
+func (s *Server) Close() {
+	if s.RunningServer == nil {
+		panic("cannot call close without first calling start")
+	}
+	s.RunningServer.Stop()
+}
+
+func (s Server) Poll(server machinist.Controller_PollServer) error {
+	for {
+		_, err := server.Recv()
+		if err != nil {
+			return err
+		}
+		response := &machinist.PollResponse{
+			Resp: &machinist.PollResponse_Pong{
+				Pong: &machinist.ActionPong{
+
+				},
+			},
+		}
+		err = server.Send(response)
+		if err != nil {
+			log.Println(err.Error())
+		}
+
+	}
+}
+
+func (s Server) Upload(server machinist.Controller_UploadServer) error {
+	panic("implement me")
+}
+
+func (s Server) Download(request *machinist.DownloadRequest, server machinist.Controller_DownloadServer) error {
+	panic("implement me")
+}
+
+type invitationToken struct {
+	Addresses  []string
+	Port       int
+	CRT        string
+	PrivateKey string
+	RootCA     string
+}
+
+func (s Server) GenerateInvitation(tags map[string]string, name string) ([]byte, error) {
+	nats, err := net.Interfaces()
```

Comment:
The set of ip addresses (or domain names) will need to come from a configuration file. If we run this in a docker container, on GCP, or even on our own internal firewall, the IP addresses of the interfaces will have nothing to do with the address configured on the network interfaces of the machines.

(eg, on GCP, it'll be a public GCP IP address, firewall will have internal 10 addresses rather than  public ones, docker runs by default - similar to k8s? - with a different IP per container, and has its own magic to remap it).

I strongly recommend breaking this down in two separate functions, one that generates the invitation, and another to ... read/compute/... the list of names/ips.